// react-tour/dist/index.js
'use strict';

var React = require('react');
var lodashThrottle = require('lodash.throttle');

var ReactTour = React.createClass({
  displayName: 'ReactTour',

  propTypes: {
    config: React.PropTypes.arrayOf(React.PropTypes.shape({
      node: React.PropTypes.object,
      text: React.PropTypes.string
    })),
    currentStep: React.PropTypes.number.isRequired,
    visible: React.PropTypes.bool,
    showDots: React.PropTypes.bool
  },

  getDefaultProps: function getDefaultProps() {
    return {
      visible: true,
      showDots: true
    };
  },

  componentDidMount: function componentDidMount() {
    this.throttledOnWindowResize = lodashThrottle(this.onWindowResize, 64);
    window.addEventListener('resize', this.throttledOnWindowResize);
  },

  componentWillUnmount: function componentWillUnmount() {
    window.removeEventListener('resize', this.throttledOnWindowResize);
  },

  onWindowResize: function onWindowResize() {
    this.forceUpdate();
  },

  render: function render() {
    var _props = this.props;
    var currentStep = _props.currentStep;
    var config = _props.config;
    var visible = _props.visible;
    var showDots = _props.showDots;

    var ElementRect;
    var classes = 'react-tour';

    if (!config) return null;
    if (!visible) classes += ' react-tour-hide';

    ElementRect = config[currentStep] ? config[currentStep].node.getBoundingClientRect() : config[0].node.getBoundingClientRect();

    return React.createElement(
      'div',
      { className: classes },
      React.createElement(
        'div',
        { className: 'react-tour-focus', style: {
            top: parseInt(ElementRect.top - 20, 10),
            left: parseInt(ElementRect.left - 20, 10),
            height: parseInt(ElementRect.height + 40, 10),
            width: parseInt(ElementRect.width + 40, 10)
          } },
        React.createElement(
          'div',
          { className: 'react-tour-focus-text' },
          config[currentStep] ? config[currentStep].text : config[0].text,
          showDots && React.createElement(
            'div',
            { className: 'react-tour-dots' },
            config.map(function (el, index) {
              return React.createElement('div', { className: 'react-tour-dots-single' + (currentStep >= index ? ' active' : ''), key: index });
            })
          )
        )
      )
    );
  }

});

module.exports = ReactTour;
